service/src/main/java/bio/terra/workspace/app/controller/WorkspaceV2Api.java

```diff
@@ -0,0 +1,218 @@
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Component;
+
+/**
+ * This class contains the implementation of the WorkspaceV2 endpoints. They are directly forwarded
+ * from the controller to here.
+ */
+@Component
+public class WorkspaceV2Api {
+  private static final Logger logger = LoggerFactory.getLogger(WorkspaceV2Api.class);
+  private final AuthenticatedUserRequestFactory authenticatedUserRequestFactory;
+  private final JobApiUtils jobApiUtils;
+  private final JobService jobService;
+  private final HttpServletRequest request;
+  private final SamService samService;
+  private final WorkspaceApiUtils workspaceApiUtils;
+  private final WorkspaceService workspaceService;
+
+  public WorkspaceV2Api(
+      AuthenticatedUserRequestFactory authenticatedUserRequestFactory,
+      JobApiUtils jobApiUtils,
+      JobService jobService,
+      HttpServletRequest request,
+      SamService samService,
+      WorkspaceApiUtils workspaceApiUtils,
+      WorkspaceService workspaceService) {
+    this.authenticatedUserRequestFactory = authenticatedUserRequestFactory;
+    this.jobApiUtils = jobApiUtils;
+    this.jobService = jobService;
+    this.request = request;
+    this.samService = samService;
+    this.workspaceApiUtils = workspaceApiUtils;
+    this.workspaceService = workspaceService;
+  }
+
+  public ResponseEntity<ApiCreateWorkspaceV2Result> createWorkspaceV2(
+      ApiCreateWorkspaceV2Request body) {
+    AuthenticatedUserRequest userRequest = authenticatedUserRequestFactory.from(request);
+    logger.info(
+        "Creating workspace {} and {} cloud context for {} subject {}",
+        body.getId(),
+        (body.getCloudPlatform() == null ? "no" : body.getCloudPlatform()),
+        userRequest.getEmail(),
+        userRequest.getSubjectId());
+    String jobId = body.getJobControl().getId();
+
+    // If we are creating a cloud context (cloudPlatform is present),
+    // then the spend profile must be valid and the user must have
+    // permission to link to it.
+    CloudPlatform cloudPlatform = null;
+    SpendProfile spendProfile = null;
+    SpendProfileId spendProfileId = null;
+    ApiCloudPlatform apiCloudPlatform = body.getCloudPlatform();
+    if (apiCloudPlatform != null) {
+      ControllerValidationUtils.validateCloudPlatform(apiCloudPlatform);
+      cloudPlatform = CloudPlatform.fromApiCloudPlatform(apiCloudPlatform);
+      spendProfile =
+          workspaceApiUtils.validateSpendProfilePermission(userRequest, body.getSpendProfile());
+      if (spendProfile == null) {
+        throw new MissingRequiredFieldsException(
+            "To create a cloud context you must provide both cloudPlatform and spendProfile");
+      }
+      spendProfileId = spendProfile.id();
+    }
+
+    TpsPolicyInputs policies =
+        workspaceApiUtils.validateAndConvertPolicies(body.getPolicies(), body.getStage());
+    // WSM requires a userFacingId. Create one, if it is not provided.
```

Comment:
Probably NA for this review since it's copied from the previous code. But this seems to be on the wrong line and I don't see where we create an id - seems like we check 2 places on the body.

Reply:
I moved the comment to the correct line.